File: csharp/3.2/PubNub-Messaging/Pubnub_Example.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.ComponentModel;

namespace PubNub_Messaging
{
    public class Pubnub_Example
    {
        static public void Main()
        {

            Console.WriteLine("\nRunning publish()");
            Publish_Example();
            
            Console.WriteLine("\nRunning history()");
            History_Example();

            Console.WriteLine("\nRunning timestamp()");
            Timestamp_Example();

            Console.WriteLine("\nRunning here_now()");
            HereNow_Example();

            Console.WriteLine("\nRunning presence()");
            Presence_Example();

            Console.WriteLine("\nRunning timestamp()");
            Subscribe_Example();

            Console.WriteLine("\nPress any key to exit when done with demo.\n\n");
            Console.ReadKey();

        }
        static void Publish_Example()
        {
            Pubnub pubnub = new Pubnub(
                    "demo",
                    "demo",
                    "",
                    false);
            string channel = "hello_world";
            string message = "Pubnub API Usage Example - Publish";

            pubnub.PropertyChanged += delegate(object sender, PropertyChangedEventArgs e)
            {
                if (e.PropertyName == "Publish")
                {
                    Console.WriteLine("\n*********** Publish Messages *********** ");
                    Console.WriteLine(
                        "Publish Success: " + ((Pubnub)sender).Publish[0].ToString() +
                        "\nPublish Info: " + ((Pubnub)sender).Publish[1].ToString()
                        );
                }
            };
            pubnub.publish(channel, message);
        }
        static void History_Example()
        {
            Pubnub pubnub = new Pubnub(
                    "demo",
                    "demo",
                    "",
                    false);
            string channel = "hello_world";
            
            pubnub.PropertyChanged += delegate(object sender, PropertyChangedEventArgs e)
            {
                if (e.PropertyName == "History")
                {
                    Console.WriteLine("\n*********** History Messages *********** ");
                    MessageFeeder(((Pubnub)sender).History);
                }
            };
            pubnub.history(channel, 10);
        }
        static void Timestamp_Example()
        {
            Pubnub pubnub = new Pubnub(
                    "demo",
                    "demo",
                    "",
                    false);
            
            pubnub.PropertyChanged += delegate(object sender, PropertyChangedEventArgs e)
            {
                if (e.PropertyName == "Time")
                {
                    Console.WriteLine("\n********** Timestamp Messages ********** ");
                    MessageFeeder(((Pubnub)sender).Time[0]);
                }
            };
            pubnub.time();
        }
        static void Subscribe_Example()
        {
            Pubnub pubnub = new Pubnub(
                    "demo",
                    "demo",
                    "",
                    false);
            string channel = "hello_world";

            pubnub.PropertyChanged += delegate(object sender, PropertyChangedEventArgs e)
            {
                if (e.PropertyName == "ReturnMessage")
                {
                    Console.WriteLine("\n********** Subscribe Messages ********** ");
                    MessageFeeder(((Pubnub)sender).ReturnMessage);
                }
            };
            pubnub.subscribe(channel);
        }

        static void Presence_Example()
        {
            Pubnub pubnub = new Pubnub(
                    "demo",
                    "demo",
                    "",
                    false);
            string channel = "hello_world";

            pubnub.PropertyChanged += delegate(object sender, PropertyChangedEventArgs e)
            {
                if (e.PropertyName == "ReturnMessage")
                {
                    Console.WriteLine("\n********** Presence Messages ********** ");
                    MessageFeeder(((Pubnub)sender).ReturnMessage);
                }
            };
            pubnub.presence(channel);
        }

        static void HereNow_Example()
        {
            Pubnub pubnub = new Pubnub(
                    "demo",
                    "demo",
                    "",
                    false);
            //string channel = "hello_world";
            string channel = "hello_world";

            pubnub.PropertyChanged += delegate(object sender, PropertyChangedEventArgs e)
            {
                if (e.PropertyName == "Here_Now")
                {
                    Console.WriteLine("\n********** Here Now Messages *********** ");
                    Dictionary<string, object> _message = (Dictionary<string, object>)(((Pubnub)sender).Here_Now[0]);
                    foreach (object uuid in (object[])_message["uuids"])
                    {
                        Console.WriteLine("UUID: " + uuid.ToString());
                    }
                    Console.WriteLine("Occupancy: " + _message["occupancy"].ToString());
                }
            };
            pubnub.here_now(channel);
        }

        static void MessageFeeder(List<object> feed)
        {
            foreach (object message in feed)
            {
                try
                {
                    Dictionary<string, object> _messageHistory = (Dictionary<string, object>)(message);
                    Console.WriteLine("Key: " + _messageHistory.ElementAt(0).Key + " - Value: " + _messageHistory.ElementAt(0).Value);
                }
                catch
                {
                    Console.WriteLine(message.ToString());
                }
            }
        }
        static void MessageFeeder(object feed)
        {
            try
            {
                Dictionary<string, object> _message = (Dictionary<string, object>)(feed);
                for (int i = 0; i < _message.Count; i ++)
                    Console.WriteLine("Key: " + _message.ElementAt(i).Key + " - Value: " + _message.ElementAt(i).Value);
            }
            catch
            {
                try
                {
                    List<object> _message = (List<object>)feed;
                    for (int i = 0; i < _message.Count; i++)
                        Console.WriteLine(_message[i].ToString());
                }
                catch
                {
                    Console.WriteLine("Time: " + feed.ToString());
                }
                
            }
        }
    }
}

File: java/3.3/src/pubnub/example/PubnubExample.java
package src.pubnub.example;

import java.util.HashMap;
import java.util.Iterator;

import org.json.JSONArray;
import org.json.JSONObject;
import pubnub.api.Pubnub;
import pubnub.api.Callback;

public class PubnubExample {

	/**
	 * @param params
	 */
	public static void main(String[] params) {
		
		System.out.println("\nRunning publish()");
		PublishExample();
		
		System.out.println("\nRunning history()");
		HistoryExample();
		
		System.out.println("\nRunning timestamp()");
		TimestampExample();
		
		System.out.println("\nRunning here_now()");
		HereNowExample();
				
		System.out.println("\nRunning detailedHistory()");
		DetailedHistoryExample();
		
//		System.out.println("\nRunning presence()");
//		PresenceExample();
		
		System.out.println("\nRunning subscribe()");
		SubscribeExample();

	}
	
	private static void PublishExample() {
		String publish_key = "demo";
		String subscribe_key = "demo";
		String secret_key = "demo";
		String cipher_key = "enigma"; // (Cipher key is optional)
		String channel = "hello_world";
		
		int publish_message_count = 1;
		
		Pubnub pubnub = new Pubnub(
				publish_key,
				subscribe_key,
				secret_key,
				cipher_key,
				true
		);
		int count = 0;
		while (true) {
			if (count >= publish_message_count) break;
			count ++;
			
			// Create JSON Message
			JSONObject message = new JSONObject();
			try {
				message.put("text", "Hello World!" + count);
				/*
				 * message.put("title", "Java Client PubNub";
				 * message.put("some_val", "This is a push to all users! Fighting!"
				 * message.put("url", "http://www.pubnub.com"
				 */
			} catch (org.json.JSONException jsonError) {
			}
			
			// Publish
			HashMap<String, Object> args = new HashMap<String, Object>(2);
			args.put("channel", channel);
			args.put("message", message);
			JSONArray response = null;
			response = pubnub.publish(args);
			System.out.println(response);
			
			args = new HashMap<String, Object>(2);
			args.put("channel", channel);
			args.put("message", "Hello World");
			response = pubnub.publish(args);
			System.out.println(response);
			
			JSONArray array = new JSONArray();
			array.put("Sunday");
			array.put("Monday");
			array.put("Tuesday");
			array.put("Wednesday");
			array.put("Thursday");
			array.put("Friday");
			array.put("Saturday");
			
			args = new HashMap<String, Object>(2);
			args.put("channel", channel);
			args.put("message", array);
			
			response = pubnub.publish(args);
			System.out.println(response);
		}
	}
	
	private static void HistoryExample() {
		String publish_key = "demo";
		String subscribe_key = "demo";
		String secret_key = "demo";
		String cipher_key = "enigma"; // (Cipher key is optional)
		String channel = "hello_world";
		int limit = 1;
		
		Pubnub pubnub = new Pubnub(
				publish_key,
				subscribe_key,
				secret_key,
				cipher_key,
				true
		);
		
		HashMap<String, Object> args = new HashMap<String, Object>(2);
		args.put("channel", channel);
		args.put("limit", limit);
		
		// Get History
		JSONArray response = pubnub.history(args);
		
		// Print Response from PubNub JSONP REST Service
		System.out.println(response);
		
		try {
			if (response != null) {
				for (int i = 0; i < response.length(); i ++) {
					JSONObject jsono = response.optJSONObject(i);
					if (jsono != null) {
						@SuppressWarnings("rawtypes")
						Iterator keys = jsono.keys();
						while (keys.hasNext()) {
							System.out.println(jsono.get(keys.next().toString()) + " ");
						}
					}
				}
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
	}
	
	private static void DetailedHistoryExample() {
		String publish_key = "demo";
		String subscribe_key = "demo";
		String secret_key = "demo";
		String cipher_key = "enigma"; // (Cipher key is optional)
		String channel = "hello_world";
		int    count = 1;
		
		Pubnub pubnub = new Pubnub(
				publish_key,
				subscribe_key,
				secret_key,
				cipher_key,
				true
		);
		
		// Get History
		JSONArray response = pubnub.detailedHistory(channel, count);
		
		// Print Response from PubNub JSONP REST Service
		System.out.println(response);
		
		try {
			if (response != null) {
				for (int i = 0; i < response.length(); i ++) {
					JSONObject jsono = response.optJSONObject(i);
					if (jsono != null) {
						@SuppressWarnings("rawtypes")
						Iterator keys = jsono.keys();
						while (keys.hasNext()) {
							System.out.println(jsono.get(keys.next().toString()) + " ");
						}
					}
				}
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
	}
	
	private static void TimestampExample() {
		String publish_key = "demo";
		String subscribe_key = "demo";
		String secret_key = "demo";
		String cipher_key = ""; // (Cipher key is optional)
		
		Pubnub pubnub = new Pubnub(
				publish_key,
				subscribe_key,
				secret_key,
				cipher_key,
				true
		);
		
		// Print Server Time
		System.out.println("Time: " + Double.toString(pubnub.time()));
	}
	
	private static void SubscribeExample() {
		String publish_key = "demo";
		String subscribe_key = "demo";
		String secret_key = "demo";
		String cipher_key = ""; // (Cipher key is optional)
		String channel = "hello_world";
		
		Pubnub pubnub = new Pubnub(
				publish_key,
				subscribe_key,
				secret_key,
				cipher_key,
				true
		);
		
		// Callback Interface when a Message is Received
        class Receiver implements Callback {

        	public boolean subscribeCallback(String channel, Object message) {

                try {
                    if (message instanceof JSONObject) {
                        JSONObject obj = (JSONObject) message;
                        @SuppressWarnings("rawtypes")
						Iterator keys = obj.keys();
                        while (keys.hasNext()) {
                            System.out.print(obj.get(keys.next().toString()) + " ");
                        }
                        System.out.println();
                    } else if (message instanceof String) {
                        String obj = (String) message;
                        System.out.print(obj + " ");
                        System.out.println();
                    } else if (message instanceof JSONArray) {
                        JSONArray obj = (JSONArray) message;
                        System.out.print(obj.toString() + " ");
                        System.out.println();
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
                // Continue Listening?
              return true;
            }

			@Override
			public void errorCallback(String channel, Object message) {
				System.err.println("Channel:" + channel + "-" + message.toString());
				
			}

			@Override
			public void connectCallback(String channel) {
				System.out.println("Connected to channel :" + channel);
				System.out.println("Waiting for a message from publisher ...");
			}

			@Override
			public void reconnectCallback(String channel) {
				System.out.println("Reconnected to channel :" + channel);
			}

			@Override
			public void disconnectCallback(String channel) {
				System.out.println("Disconnected to channel :" + channel);
			}

			@Override
			public boolean presenceCallback(String channel, Object message) {
				return false;
			}
        }

        HashMap<String, Object> args = new HashMap<String, Object>(6);
        args.put("channel", channel);
        args.put("callback", new Receiver());					// callback to get response

        // Listen for Messages (Subscribe)
        pubnub.subscribe(args);
	}
	
	private static void PresenceExample() {
		String publish_key = "demo";
		String subscribe_key = "demo";
		String secret_key = "demo";
		String cipher_key = ""; // (Cipher key is optional)
		String channel = "hello_world";
		
		Pubnub pubnub = new Pubnub(
				publish_key,
				subscribe_key,
				secret_key,
				cipher_key,
				true
		);
		
		// Callback Interface when a Message is Received
        class Receiver implements Callback {

        	public boolean presenceCallback(String channel, Object message) {

                try {
                    if (message instanceof JSONObject) {
                        JSONObject obj = (JSONObject) message;
                        @SuppressWarnings("rawtypes")
						Iterator keys = obj.keys();
                        while (keys.hasNext()) {
                            System.out.print(obj.get(keys.next().toString()) + " ");
                        }
                        System.out.println();
                    } else if (message instanceof String) {
                        String obj = (String) message;
                        System.out.print(obj + " ");
                        System.out.println();
                    } else if (message instanceof JSONArray) {
                        JSONArray obj = (JSONArray) message;
                        System.out.print(obj.toString() + " ");
                        System.out.println();
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
                // Continue Listening?
              return true;
            }

			@Override
			public void errorCallback(String channel, Object message) {
				System.err.println("Channel:" + channel + "-" + message.toString());
				
			}

			@Override
			public void connectCallback(String channel) {
				System.out.println("Connected to channel :" + channel);
				System.out.println("Waiting for subscribe or unsubscribe message ...");
			}

			@Override
			public void reconnectCallback(String channel) {
				System.out.println("Reconnected to channel :" + channel);
			}

			@Override
			public void disconnectCallback(String channel) {
				System.out.println("Disconnected to channel :" + channel);
			}

			@Override
			public boolean subscribeCallback(String channel, Object message) {
				return false;
			}
        }

        HashMap<String, Object> args = new HashMap<String, Object>(6);
        args.put("channel", channel + "-pnpres");
        args.put("callback", new Receiver());					// callback to get response

        // Listen for Messages (Presence)
        pubnub.presence(args);
	}
	
	private static void HereNowExample() {
		String publish_key = "demo";
		String subscribe_key = "demo";
		String secret_key = "demo";
		String cipher_key = ""; // (Cipher key is optional)
		String channel = "hello_world";
		
		Pubnub pubnub = new Pubnub(
				publish_key,
				subscribe_key,
				secret_key,
				cipher_key,
				true
		);
		
		// Get Here Now
		JSONArray response = pubnub.here_now(channel);
		
		// Print Response from PubNub JSONP REST Service
		System.out.println(response);
		
		try {
			if (response != null) {
				for (int i = 0; i < response.length(); i ++) {
					JSONObject jsono = response.optJSONObject(i);
					if (jsono != null) {
						@SuppressWarnings("rawtypes")
						Iterator keys = jsono.keys();
						while (keys.hasNext()) {
							System.out.println(jsono.get(keys.next().toString()) + " ");
						}
					}
				}
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
	}
}
